### PR TITLE
Add table component

### PR DESCRIPTION
**Pull request by maxbischof**
Table component renders csv data as a table.
See [User Story](https://github.com/maxbischof/data-hub-nord/issues/4) for further Information

### CODE REVIEW

File: src/stories/Button.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
```
Comment (constantingoeldel) [inline src/stories/Button.js:1]: Te storybook examples should are not necessary on GitHub, Jan posted a way to accomplish this: 
1. `$ git rm -r --cached <path-to-storybook-examples>` 
This will remove the folder from the index, but keep your local copy. If you are unsure about this use the --dry-run flag to get a preview of what will be removed.
2. `Add <path-to-storybook-examples>` to your .gitignore
3. Commit and push again

File: src/stories/Table.stories.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
```
Comment (constantingoeldel) [inline src/stories/Table.stories.js:1]: If you use the example Array to render the table in this file, what do you need the src/stories/Table.js file for? Which one is actually necessary?

Comment (maxbischof) [reply src/stories/Table.stories.js:1]: Forgot to remove src/stories/Table.js. It's unnecessary. Thanks!

File: src/components/Table.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import React from "react"
+import PropTypes from "prop-types"
+import { uid } from "react-uid"
+import styled from 'styled-components'
+
+export default function Table({ array }) {
```
Comment (constantingoeldel) [inline src/components/Table.js:6]: I don't think array is a good prop name, as it is generic and not informative
What about `tableData`, for example?

**Review by constantingoeldel — APPROVED** [review]

File: src/stories/Button.js (hunk 1 above)
Comment (maxbischof) [reply src/stories/Button.js:1]: Thanks for the hint.

File: .storybook/preview.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React from 'react'
+import GlobalStyles from '../src/GlobalStyles'
+
+export const decorators = [
+  (Story) => (
+    <React.Fragment>
```
Comment (badabam) [inline .storybook/preview.js:6]: Please consider using `<>` and `</>`

File: src/App.js [hunk 5]
```diff
@@ -1,30 +1,25 @@
...
+import React, { useState, useEffect } from "react"
 import { datasets } from "./settings.js"
 import { fetchCSV, csvToObjectsArray } from "./utils.js"
+import Table from './components/Table'
 
 function App() {
-  fetchCSV({ path: datasets[0].url }).then((response) =>
-    console.log(
-      csvToObjectsArray({
+  const [dataset1, setDataset1] = useState([{}])
```
Comment (badabam) [inline src/App.js:7]: Why empty object in the array? Maybe find a better name for `dataset`.

File: src/App.js [hunk 6]
```diff
@@ -1,30 +1,25 @@
...
+
+  useEffect(() => {
+    fetchCSV({ path: datasets[0].url }).then((response) => {
+      const array = csvToObjectsArray({
```
Comment (badabam) [inline src/App.js:11]: `array` is maybe not the best name for a variable – I know it's hard in your case to find better wordings.

File: src/App.js [hunk 7]
```diff
@@ -1,30 +1,25 @@
...
-        seperator: datasets[1].seperator,
-      })
-    )
-  )
+      setDataset1(array)
+    })
+  }, [])
 
-  return <div className="App"></div>
+  return <div className="App">
```
Comment (badabam) [inline src/App.js:20]: Is there a class `App` in your styles?

File: src/components/Table.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+              <tr>
+                {columnNames.map((name) => (
+                  <td key={name}>{name}</td>
+                ))}
+              </tr>
+            </TableHeader>
+
+            <TableBody>
+              {array.map((object) => (
+                <tr key={uid(object)}>
```
Comment (badabam) [inline src/components/Table.js:23]: Nice idea. The uids are created on every render, though. Consider adding them once initially.

File: src/components/Table.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                  ))}
+                </tr>
+              ))}
+            </TableBody>
+          </StyledTable>
+        </OverflowContainer>
+  )
+}
+
+Table.propTypes = {
```
Comment (badabam) [inline src/components/Table.js:35]: Yay for propTypes!

File: src/components/Table.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+const OverflowContainer = styled.div`
+  overflow-x: scroll;
+`
+
+const StyledTable = styled.table`
+  border-spacing: 0;
+
+  td{
```
Comment (badabam) [inline src/components/Table.js:46]: Missing space.

File: src/components/Table.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    padding: 10px;
+  }
+
+  tr > td:first-child{
```
Comment (badabam) [inline src/components/Table.js:50]: Missing space.

File: src/components/Table.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    position: sticky;
```
Comment (badabam) [inline src/components/Table.js:51]: As people often mistake `sticky` for `fixed` – do you really want this to scroll until it is on the left side? Or always fixed?

File: src/components/Table.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    top: 0;
+    width: 5px;
+  }
+`
+
+const TableHeader = styled.thead`
+  color: white;
+
+  td{
+    border: 0px solid #1D3557;
```
Comment (badabam) [inline src/components/Table.js:70]: Just `0`, not `0px`

File: src/components/Table.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  tr:nth-child(2n) > td:first-child{
+    background: #EFF0F0;
+  }
+
+  tr:nth-child(2n) {
+    background: #EFF0F0;
+  }
+`
```
Comment (badabam) [inline src/components/Table.js:102]: Please configure your Editor to add a newline at the end.

File: src/components/Table.test.js [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+  de-sh;Kiel;wirtschaft_arbeit;Unterbesch�ftigung;2018;16608;11,5
+  de-sh;Kiel;wirtschaft_arbeit;Unterbesch�ftigung;2019;15453;10,8
+  `
+const exampleArray = csvToObjectsArray({ csv: exampleCSV, columnNames: [], seperator: ";" })
+
+it('renders table component', () => {
+  const renderedTable = renderer
+    .create(<Table array={exampleArray}></Table>)
+    .toJSON();
+  expect(renderedTable).toMatchSnapshot();
```
Comment (badabam) [inline src/components/Table.test.js:25]: Nice!

File: src/components/Table.test.js [hunk 16]
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import Table from './Table';
+import renderer from 'react-test-renderer';
+import { csvToObjectsArray } from '../utils';
+
+const exampleCSV = `Land;Stadt;Kategorie;Merkmal;Jahr;Unterbsch�ftigte;Unterbsch�ftigtenquote
```
Comment (badabam) [inline src/components/Table.test.js:6]: There seems something wrong with the umlauts

File: src/stories/Button.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import PropTypes from 'prop-types';
+import './button.css';
+
+/**
+ * Primary UI component for user interaction
+ */
+export const Button = ({ primary, backgroundColor, size, label, ...props }) => {
+  const mode = primary ? 'storybook-button--primary' : 'storybook-button--secondary';
```
Comment (badabam) [inline src/stories/Button.js:9]: What is this button? Is it only for storybook?

File: src/stories/Button.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+  return (
+    <button
+      type="button"
+      className={['storybook-button', `storybook-button--${size}`, mode].join(' ')}
+      style={backgroundColor && { backgroundColor }}
```
Comment (badabam) [inline src/stories/Button.js:14]: Inline styles?

File: src/stories/Button.stories.js [hunk 19]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+
+import { Button } from './Button';
+
+export default {
+  title: 'Example/Button',
```
Comment (badabam) [inline src/stories/Button.stories.js:6]: This is an example story? Maybe delete this and the weird Button from above?

File: src/stories/Header.js [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+import PropTypes from 'prop-types';
+
+import { Button } from './Button';
+import './header.css';
+
+export const Header = ({ user, onLogin, onLogout, onCreateAccount }) => (
+  <header>
+    <div className="wrapper">
+      <div>
+        <svg width="32" height="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg">
```
Comment (badabam) [inline src/stories/Header.js:11]: You can extract svgs in normal svg files and import them. e.g. `import {ReactComponent as MyIcon} from '../images/myicon.svg'`

Comment (badabam) [reply src/stories/Header.js:11]: Then use it like this: `<MyIcon />`

File: src/stories/Header.js [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+              d="M27.2 10.6v11.2l-10.5 6V16.5l10.5-6zM15.7 4.4v11L6 10l9.7-5.5z"
+              fill="#91BAF8"
+            />
+          </g>
+        </svg>
+        <h1>Acme</h1>
+      </div>
+      <div>
+        {user ? (
+          <Button size="small" onClick={onLogout} label="Log out" />
```
Comment (badabam) [inline src/stories/Header.js:31]: Consider using the `children` prop instead of `label`.

File: src/stories/Header.js [hunk 22]
```diff
@@ -0,0 +1,52 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import { Button } from './Button';
+import './header.css';
+
+export const Header = ({ user, onLogin, onLogout, onCreateAccount }) => (
```
Comment (badabam) [inline src/stories/Header.js:7]: You already have a login in your app? Do you have a server and user management in place? Otherwise this should not be here.

File: src/stories/Introduction.stories.mdx [hunk 23]
```diff
@@ -0,0 +1,194 @@
+import { Meta } from '@storybook/addon-docs/blocks';
```
Comment (badabam) [inline src/stories/Introduction.stories.mdx:1]: Please delete this file.

File: src/stories/Page.js [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+
+import { Header } from './Header';
+import './page.css';
+
+export const Page = ({ user, onLogin, onLogout, onCreateAccount }) => (
+  <article>
+    <Header user={user} onLogin={onLogin} onLogout={onLogout} onCreateAccount={onCreateAccount} />
+
+    <section>
+      <h2>Pages in Storybook</h2>
```
Comment (badabam) [inline src/stories/Page.js:12]: Why is this here?

File: src/stories/Page.stories.js [hunk 25]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+import { Page } from './Page';
+import * as HeaderStories from './Header.stories';
+
+export default {
+  title: 'Example/Page',
```
Comment (badabam) [inline src/stories/Page.stories.js:7]: Please remove those examples.

File: src/stories/button.css [hunk 26]
```diff
@@ -0,0 +1,30 @@
+.storybook-button {
```
Comment (badabam) [inline src/stories/button.css:1]: That's not yours, right? Please delete this.

File: src/stories/header.css [hunk 27]
```diff
@@ -0,0 +1,26 @@
+.wrapper {
```
Comment (badabam) [inline src/stories/header.css:1]: Can be deleted?

File: src/stories/page.css [hunk 28]
```diff
@@ -0,0 +1,69 @@
+section {
```
Comment (badabam) [inline src/stories/page.css:1]: Can be deleted?

File: src/components/Table.js (hunk 3 above)
Comment (maxbischof) [reply src/components/Table.js:6]: It's not easy because the table component is generic itself. TableData is better than array but still I am not happy with it. Changed it to tableData for now.

File: package.json [hunk 29]
```diff
@@ -8,13 +8,18 @@
     "@testing-library/user-event": "^7.2.1",
     "react": "^16.13.1",
     "react-dom": "^16.13.1",
-    "react-scripts": "3.4.3"
+    "react-scripts": "3.4.3",
+    "react-test-renderer": "^16.13.1",
```
[the remark this answers is not in the document]
Comment (maxbischof) [reply package.json:12]: Moved them to devDependencies.

File: src/App.js (hunk 5 above)
Comment (maxbischof) [reply src/App.js]: The empty object was workaround for not validating if this data is present or not. I added validation in the table component with `if(tableData.length > 0)` and removed the empty object.

```javascript
export default function Table({ tableData }) {
  let headerRow = null
  let bodyRows = null

  if(tableData.length > 0) {
    const columnNames = Object.keys(tableData[0])

    headerRow = (
      columnNames.map((name) => (
        <td key={name}>{name}</td>
      ))
    )

    bodyRows = (
      tableData.map((object) => (
        <tr key={uid(object)}>
          {columnNames.map((name) => (
            <td key={uid(name)}>{object[name]}</td>
          ))}
        </tr>
      ))
    )
  }

  return (
        <OverflowContainer>
          <HeaderBackground />
          <StyledTable>
            <TableHeader>
              <tr>
                {headerRow}
              </tr>
            </TableHeader>

            <TableBody>
              {bodyRows}
            </TableBody>
          </StyledTable>
        </OverflowContainer>
  )
}
```

File: src/App.js (hunk 6 above)
Comment (maxbischof) [reply src/App.js:11]: I changed it to `tableData`. At least its more specific. But still it feels like a bad naming.

File: src/stories/Page.js (hunk 24 above)
Comment (maxbischof) [reply src/stories/Page.js:12]: Storybook example data. Constantin already suggested to remove it from the repo.

File: src/App.js (hunk 5 above)
Comment (badabam) [reply src/App.js]: `if (data.length)` would be enough for me. It's already boolish (ie. truthy/falsy).

File: src/components/Table.test.js (hunk 16 above)
Comment (maxbischof) [reply src/components/Table.test.js:6]: The csv is iso-8859-1 encoded. So the default of utf-8 was wrong. Now I am decoding the response with the right charset.
